pkg/agent/proxy/proxier.go

```diff
@@ -542,35 +559,47 @@ func (p *proxier) installServices() {
 			}
 
 			// Install ClusterIP flows for the Service.
-			groupID := p.groupCounter.AllocateIfNotExist(svcPortName, internalPolicyLocal)
+			groupID, exists := p.groupCounter.Get(svcPortName, internalPolicyLocal)
+			if !exists {
+				klog.ErrorS(nil, "Group for Service internalTrafficPolicy was not installed", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr, "internalTrafficPolicy", internalPolicyLocal)
+				continue
+			}
 			if err := p.ofClient.InstallServiceFlows(groupID, svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol, uint16(affinityTimeout), externalPolicyLocal, corev1.ServiceTypeClusterIP, isNestedService); err != nil {
 				klog.Errorf("Error when installing Service flows: %v", err)
 				continue
 			}
 
 			if p.proxyAll {
-				nGroupID := p.groupCounter.AllocateIfNotExist(svcPortName, externalPolicyLocal)
 				// Install ClusterIP route on Node so that ClusterIP can be accessed on Node. Every time a new ClusterIP
 				// is created, the routing target IP block will be recalculated for expansion to be able to route the new
 				// created ClusterIP. Deleting a ClusterIP will not shrink the target routing IP block. The Service CIDR
 				// can be finally calculated after creating enough ClusterIPs.
 				if err := p.routeClient.AddClusterIPRoute(svcInfo.ClusterIP()); err != nil {
-					klog.ErrorS(err, "Failed to install ClusterIP route of Service", "Service", svcPortName)
+					klog.ErrorS(err, "Error when installing ClusterIP route for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
 					continue
 				}
 
 				// If previous Service is nil or NodePort flows and configurations of previous Service have been removed,
 				// install NodePort flows and configurations for current Service.
 				if svcInfo.NodePort() > 0 && (pSvcInfo == nil || needRemoval) {
-					if err := p.installNodePortService(nGroupID, uint16(svcInfo.NodePort()), svcInfo.OFProtocol, uint16(affinityTimeout), svcInfo.ExternalPolicyLocal()); err != nil {
-						klog.ErrorS(err, "Failed to install NodePort flows and configurations of Service", "Service", svcPortName)
+					groupID, exists = p.groupCounter.Get(svcPortName, externalPolicyLocal)
```

Comment:
Question: why do you expect the OF group for the NodePort traffic to share the same identifier as the ClusterIP traffic for a Service? In the previous implementation, here calls `p.groupCounter.AllocateIfNotExist` to get or create an OF group ID for the new configured NodePort scenario.

Reply:
>  In the previous implementation, here calls p.groupCounter.AllocateIfNotExist to get or create an OF group ID for the new configured NodePort scenario.

We should use method `Get` here to get a OVS group ID, otherwise, there is a chance that we allocate a group ID and pass it to `InstallServiceGroup` and the corresponding group was not created in OVS.